src/Haskell/Ide/Engine/Transport/LspStdio.hs

```diff
@@ -98,33 +123,49 @@ run
   -> IO Int
 run dispatcherProc cin _origDir plugins captureFp = flip E.catches handlers $ do
 
-  rin <- atomically newTChan :: IO (TChan ReactorInput)
+  rin        <- atomically newTChan :: IO (TChan ReactorInput)
   commandIds <- allLspCmdIds plugins
 
   let dp lf = do
         cancelTVar  <- atomically $ newTVar S.empty
         wipTVar     <- atomically $ newTVar S.empty
         versionTVar <- atomically $ newTVar Map.empty
+        diagIn      <- atomically newTChan
         let dEnv = DispatcherEnv
               { cancelReqsTVar = cancelTVar
               , wipReqsTVar    = wipTVar
               , docVersionTVar = versionTVar
               }
-        let reactorFunc = runReactor lf dEnv cin diagnosticProviders hps sps
-              $ reactor rin
+        let react = runReactor lf dEnv cin diagnosticProviders hps sps
+        let reactorFunc = react $ reactor rin diagIn
             caps = Core.clientCapabilities lf
 
         let errorHandler :: ErrorHandler
             errorHandler lid code e =
               Core.sendErrorResponseS (Core.sendFunc lf) (J.responseId lid) code e
             callbackHandler :: CallbackHandler R
-            callbackHandler f x = runReactor lf dEnv cin diagnosticProviders hps sps $ f x
+            callbackHandler f x = react $ f x
+
+        -- This is the callback the debouncer executes at the end of the timeout,
+        -- it executes the diagnostics for the most recent request.
+        let dispatchDiagnostics :: Option (MostRecent DiagnosticsRequest) -> R ()
+            dispatchDiagnostics req = option (pure ()) (requestDiagnostics . coerce) req
+
+        -- Debounces messages published to the diagnostics channel.
+        let diagnosticsQueue tr = forever $ do
+              inval <- liftIO $ atomically $ readTChan diagIn
+              Debounce.send tr (coerce . Just $ MostRecent inval)
 
+        tr <- Debounce.new -- Debounce for 350ms
+          (Debounce.forMonoid $ react . dispatchDiagnostics)
```

Comment:
I can't seem to get rid of the need for running this in the `R` monad. The reason is that `requestDiagnostics` lives there, and I need to call that function from inside the Debounce callback.

There are some gymnastics I can make, like duplicating the code and creating a separate `REnv` that I can pass as a parameter to `requestDiagnostics`, but that seems a lot dirtier than doing this. I any suggestions @alanz @bubba ?

Reply:
Theoretically a reader environment should be safe.  The complications happen when you have shared vars in it.  Which are things I have mixed feelings about at the best of times.

Reply:
I'm not quite sure why `dispatcherEnv` is even inside `REnv`. I'm going to try pulling it out, see what happens

Reply:
Looks like they're very tightly coupled. There's a lot of dispatch related stuff like cancelling requests etc. going on in LspStdio. I think we should maybe try moving that stuff into Dispatcher so we can hide the `TVars`.

Reply:
Can we do that as a separate task and not mix it with this PR?

Reply:
I am happy to merge it as is, provided we make an issue for the cleanup.

@bubba ?